pkg/cmd/unjoin/exec.go

```diff
@@ -0,0 +1,366 @@
+// Copyright Contributors to the Open Cluster Management project
+package unjoin
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"os/user"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	k8syaml "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapiv1 "k8s.io/client-go/tools/clientcmd/api/v1"
+
+	"open-cluster-management.io/clusteradm/pkg/cmd/join/scenario"
+	"open-cluster-management.io/clusteradm/pkg/helpers"
+	"open-cluster-management.io/clusteradm/pkg/helpers/apply"
+
+	"github.com/ghodss/yaml"
+	"github.com/spf13/cobra"
+)
+
+func (o *Options) complete(cmd *cobra.Command, args []string) (err error) {
+	klog.V(1).InfoS("unjoin  options:", "dry-run", o.ClusteradmFlags.DryRun, "cluster", o.clusterName, "api-server", o.hubAPIServer, o.outputFile)
+
+	o.values = Values{
+		ClusterName: o.clusterName,
+		Hub: Hub{
+			APIServer: o.hubAPIServer,
+		},
+	}
+	klog.V(3).InfoS("values:", "clusterName", o.values.ClusterName, "hubAPIServer", o.values.Hub.APIServer)
+	return nil
+}
+
+func (o *Options) validate() error {
+	if o.token == "" {
+		return fmt.Errorf("token is missing")
+	}
+	if o.values.Hub.APIServer == "" {
+		return fmt.Errorf("hub-server is misisng")
+	}
+	if o.values.ClusterName == "" {
+		return fmt.Errorf("name is missing")
+	}
+
+	return nil
+}
+
+func (o *Options) run() error {
```

Comment:
I think your step should be

1. delete klusterlet
2. delete operator related resources when the  step 1 is done

Reply:
@qiujian16 could help review my latest changes, Thanks

Reply:
It does not seem to resolve the comment. My suggestion is:

1. unjoin should not call hub apiserver. unjoin should only delete klusterlet CR and operator on spoke with an output message to tell user that the agent is removed on spoke, user should opt to remove managedcluster on hub.
2. It might be worth to check if there are existing AppliedManifestWork on spoke, and unjoin should fail with an warning message that if user uninstall the klusterlet, some manifestworks on hub cannot be cleaned.

Reply:
@qiujian16 ， Let me explain this more clearly:

1. In my initial commit, I want to call hub apiserver to delete mcl resource, currently I have removed that.  However , time  I just wish to delete the resource `Secret bootstrap-hub-kubeconfig` which defined in `join/bootstrap_hub_kubeconfig.yaml`, and it is needed `.Hub.KubeConfig`: https://github.com/open-cluster-management-io/clusteradm/blob/d0cb45a8720243b76526fb50776b55a0c9c7997d/pkg/cmd/join/scenario/join/bootstrap_hub_kubeconfig.yaml#L9 
according to your latest [comments](https://github.com/open-cluster-management-io/clusteradm/pull/67/files#r716293511) , I don't need to remove `join/bootstrap_hub_kubeconfig.yaml` , right ? if so, I can remove calling hub apiserver actions. 

2. Yea, it is worth to add the condition to check the `AppliedManifestWork` before the user uninstall klusterlet.

Reply:
bootstrap is in the agent ns, and when klusterlet is deleted, the ns will be deleted as a whole.